src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1072,17 +1049,25 @@ static UCS_CLASS_CLEANUP_FUNC(uct_dc_mlx5_ep_t)
                        "iface (%p) ep (%p) dci leak detected: dci=%d", iface,
                        self, self->dci);
 
-    uct_dc_mlx5_txqp_purge_outstanding(iface, &iface->tx.dcis[self->dci].txqp,
-                                       UCS_ERR_CANCELED,
-                                       iface->tx.dcis[self->dci].txwq.sw_pi, 1);
+    /* TODO should be removed by flush */
+    uct_rc_txqp_purge_outstanding(&iface->super.super,
+                                  &iface->tx.dcis[self->dci].txqp, UCS_ERR_CANCELED,
+                                  iface->tx.dcis[self->dci].txwq.sw_pi, 1);
     ucs_assert(ucs_queue_is_empty(&iface->tx.dcis[self->dci].txqp.outstanding));
     iface->tx.dcis[self->dci].ep = NULL;
+
+    it = kh_get(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, (uint64_t)self);
+    if (it != kh_end(&iface->tx.fc_hash)) {
+        kh_del(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, it);
+        uct_dc_mlx5_ep_clear_fc_grant_flag(iface, self);
```

Comment:
can we remove the grant flag altogether?

Reply:
AFAIU than we won't know we've already sent request for the EP and will keep sending

Reply:
can we check in hash table?

Reply:
We can, but couldn't that affect performance?

Reply:
is it done only when we want to send grant, and already do kh_put to add to table? maybe check kh_put result to see if key already exists?